### ISSUE TITLE
Collect in a simple query fails: `Should not reach await until tracking is complete, cancelled or out-of demand`

### ISSUE THREAD

**Issue by philastrophist** [neo4j/neo4j#12657]
Neo4j raises a weird error that I can't google when collecting nodes in a very simple query

- Neo4j version: 4.2.0
- Operating system: Windows 10 - 19041.630
- API/Driver: Cypher - Neo4jDesktop

# Steps to reproduce:

```
match (ob: OB) 
optional match (ob)-->(e:Exposure) 
with ob, collect(e) as exposures
return ob, exposures
```

# Expected behavior
Returns rows of `ob` and `exposure` lists

# Actual behavior 
Raises error: 
`Neo.DatabaseError.Statement.ExecutionFailed`
`Should not reach await until tracking is complete, cancelled or out-of demand,
count: 1, cancelled: false, demand: 50`

I can return `ob` only and then use the neo4j desktop to expand the relationships myself but the entire query fails.

![Untitled](https://user-images.githubusercontent.com/14169057/103769201-be2ea880-501b-11eb-9305-ea4c2eb62faf.png)

# Other behaviour
I can't reproduce this behaviour by creating a minimalistic example.

If I run a creation script, which emulates the same type of data structure

```
WITH * unwind range(0, 3) as i
	 CREATE (a:A)
    WITH * unwind range(0,10) as j
      CREATE (b:B)
      CREATE (a)-[:rel]->(b)
RETURN a, b
```

Then I can run the query successfully
```
match (a: A) 
optional match (a)-->(b:B) 
with a, collect(b) as bs
return a, bs
```
![Untitled2](https://user-images.githubusercontent.com/14169057/103769265-e74f3900-501b-11eb-81d8-0c49348df69f.png)

This issue persists even when I delete the database and write everything back again

**Comment by philastrophist** [comment]
`neo4j.log` reads: `2021-01-06 12:40:49.567+0000 ERROR Client triggered an unexpected error [Neo.DatabaseError.Statement.ExecutionFailed]: Should not reach await until tracking is complete, cancelled or out-of demand,
count: 1, cancelled: false, demand: 50, reference 894d6c8f-62b7-40bd-9ea1-4073c41d1c60.`

`debug.log` reads:
```
2021-01-06 12:40:49.567+0000 ERROR [o.n.b.r.s.i.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.Statement.ExecutionFailed]: Should not reach await until tracking is complete, cancelled or out-of demand,
count: 1, cancelled: false, demand: 50, reference 894d6c8f-62b7-40bd-9ea1-4073c41d1c60.
2021-01-06 12:40:49.568+0000 ERROR [o.n.b.r.s.i.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.Statement.ExecutionFailed]: Should not reach await until tracking is complete, cancelled or out-of demand,
count: 1, cancelled: false, demand: 50, reference 894d6c8f-62b7-40bd-9ea1-4073c41d1c60.
org.neo4j.fabric.executor.FabricException: Should not reach await until tracking is complete, cancelled or out-of demand,
count: 1, cancelled: false, demand: 50
	at org.neo4j.fabric.executor.Exceptions.transform(Exceptions.java:51) ~[neo4j-fabric-4.2.0.jar:4.2.0]
	at org.neo4j.fabric.stream.Rx2SyncStream.readRecord(Rx2SyncStream.java:69) ~[neo4j-fabric-4.2.0.jar:4.2.0]
	at org.neo4j.fabric.bolt.BoltQueryExecutionImpl$QueryExecutionImpl.request(BoltQueryExecutionImpl.java:179) ~[neo4j-fabric-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.AbstractCypherAdapterStream.handleRecords(AbstractCypherAdapterStream.java:104) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.v3.messaging.ResultHandler.onPullRecords(ResultHandler.java:41) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.v4.messaging.PullResultConsumer.consume(PullResultConsumer.java:42) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State.consumeResult(TransactionStateMachine.java:511) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State$2.streamResult(TransactionStateMachine.java:355) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine.streamResult(TransactionStateMachine.java:94) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.v4.runtime.InTransactionState.processStreamResultMessage(InTransactionState.java:73) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.v4.runtime.AbstractStreamingState.processUnsafe(AbstractStreamingState.java:49) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.v4.runtime.InTransactionState.processUnsafe(InTransactionState.java:60) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.v3.runtime.FailSafeBoltStateMachineState.process(FailSafeBoltStateMachineState.java:48) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.nextState(AbstractBoltStateMachine.java:143) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.process(AbstractBoltStateMachine.java:91) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:90) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.lambda$enqueue$0(DefaultBoltConnection.java:147) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatchInternal(DefaultBoltConnection.java:236) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:171) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:161) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:246) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$3(ExecutorBoltScheduler.java:229) ~[neo4j-bolt-4.2.0.jar:4.2.0]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-all-4.1.53.Final.jar:4.1.53.Final]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.IllegalStateException: Should not reach await until tracking is complete, cancelled or out-of demand,
count: 1, cancelled: false, demand: 50
	at org.neo4j.cypher.internal.runtime.pipelined.state.StandardQueryCompletionTracker.await(QueryCompletionTracker.scala:226) ~[neo4j-cypher-pipelined-runtime-4.2.0.jar:4.2.0]
	at org.neo4j.cypher.internal.runtime.pipelined.execution.ExecutingQuery.await(ExecutingQuery.scala:35) ~[neo4j-cypher-pipelined-runtime-4.2.0.jar:4.2.0]
	at org.neo4j.cypher.internal.runtime.pipelined.execution.CallingThreadExecutingQuery.await(CallingThreadExecutingQuery.scala:75) ~[neo4j-cypher-pipelined-runtime-4.2.0.jar:4.2.0]
	at org.neo4j.cypher.internal.PipelinedRuntime$PipelinedRuntimeResult.await(PipelinedRuntime.scala:447) ~[neo4j-enterprise-cypher-4.2.0.jar:4.2.0]
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.await(StandardInternalExecutionResult.scala:95) ~[neo4j-cypher-4.2.0.jar:4.2.0]
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.await(ClosingExecutionResult.scala:162) ~[neo4j-cypher-4.2.0.jar:4.2.0]
	at org.neo4j.fabric.stream.QuerySubject$BasicQuerySubject$1.doRequest(QuerySubject.java:191) ~[neo4j-fabric-4.2.0.jar:4.2.0]
	at org.neo4j.fabric.stream.QuerySubject$BasicQuerySubject$1.request(QuerySubject.java:167) ~[neo4j-fabric-4.2.0.jar:4.2.0]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:130) ~[reactor-core-3.3.9.RELEASE.jar:3.3.9.RELEASE]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:130) ~[reactor-core-3.3.9.RELEASE.jar:3.3.9.RELEASE]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:130) ~[reactor-core-3.3.9.RELEASE.jar:3.3.9.RELEASE]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:130) ~[reactor-core-3.3.9.RELEASE.jar:3.3.9.RELEASE]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:130) ~[reactor-core-3.3.9.RELEASE.jar:3.3.9.RELEASE]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.request(Operators.java:2116) ~[reactor-core-3.3.9.RELEASE.jar:3.3.9.RELEASE]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:130) ~[reactor-core-3.3.9.RELEASE.jar:3.3.9.RELEASE]
	at reactor.core.publisher.StrictSubscriber.request(StrictSubscriber.java:138) ~[reactor-core-3.3.9.RELEASE.jar:3.3.9.RELEASE]
	at org.neo4j.fabric.stream.Rx2SyncStream$RecordSubscriber.request(Rx2SyncStream.java:129) ~[neo4j-fabric-4.2.0.jar:4.2.0]
	at org.neo4j.fabric.stream.Rx2SyncStream.maybeRequest(Rx2SyncStream.java:91) ~[neo4j-fabric-4.2.0.jar:4.2.0]
	at org.neo4j.fabric.stream.Rx2SyncStream.readRecord(Rx2SyncStream.java:50) ~[neo4j-fabric-4.2.0.jar:4.2.0]
	... 25 more
```

**Comment by philastrophist** [comment]
New discovery:

I can introduce a `limit` of any size I like and the query will run.

RUNS:
```
match (ob: OB) 
with ob
limit 10000000  // there are only 6 ob in the database so this does nothing
optional match (ob)-[:is_required_by]->(e:Exposure) 
with ob, collect(e) as exposures
return ob, exposures
```

If I remove the `limit`, the query produces the same error

ERRORS:
```
match (ob: OB) 
with ob
optional match (ob)-[:is_required_by]->(e:Exposure) 

with ob, collect(e) as exposures
return ob, exposures
```

So it looks like it's a planner problem. And indeed, the planner produces different strategies when using a limit:

## No limit
![nolimit](https://user-images.githubusercontent.com/14169057/103800468-6193b380-5044-11eb-8723-54e7846ec7cd.png)

## With a limit
![withlimit](https://user-images.githubusercontent.com/14169057/103800473-622c4a00-5044-11eb-9ae7-03e1bc537a82.png)


We can stop the problem from arising by using `runtime=slotted`. But this is not a solution...

**Comment by mnd999** [comment]
Thanks for the report, it certainly looks like a bug at first glance. We'll take a look and let you know what we find.

**Comment by rjbaucells** [comment]
Same error here but not related to `COLLECT()`:

```
MATCH (_2:A)-[:HAS_B]->(_3:B) 
MATCH (_1:C) WHERE (_1.unique = _2.c) 

OPTIONAL MATCH (_1)<-[m:HAS_C]-(:D)

RETURN 
    _1.number,
    sum(m.weight)
```

Removing the `OPTIONAL` keyword avoids the error, but not a solution in our case.

Failed to replicate it in an empty database with only the data traversed by the query.

Neo4J: 4.2.2 enterprise

**Comment by ToeFungi** [comment]
Seem to be experiencing the same issue. Any further information on this?

**Comment by ToeFungi** [comment]
I have to almost identical queries, the only difference being a filter parameter, the one query works while the second doesn't.

> Should not reach await until tracking is complete, cancelled or out-of demand,
count: 1, cancelled: false, demand: 1000

```
MATCH (member:Person {
  mobileNumber: $mobileNumber // Doesn't work
})-[:MEMBER_OF|VISITOR_OF|SUPER_ADMIN_OF]->(org:Org {
  orgId: $orgId
})

...
```

```
MATCH (member:Person {
  uuid: $uuid // Works for some reason
})-[:MEMBER_OF|VISITOR_OF|SUPER_ADMIN_OF]->(org:Org {
  orgId: $orgId
})

...
```

**Comment by mnd999** [comment]
A fixed has been developed for this issue, and is in the process of being merged. It will be included in the next 4.2.x release. Until then, as a workaround prefixing queries with `CYPHER runtime=slotted` should avoid the issue.
